pygfx/utils/transform.py

```diff
@@ -0,0 +1,335 @@
+
+        return cache[1]
+
+
+class AffineBase:
+    last_modified: int
+
+    @property
+    def matrix(self):
+        raise NotImplementedError()
+
+    @cached
+    def inverse_matrix(self) -> np.ndarray:
+        return np.linalg.inv(self.matrix)
+
+    @cached
+    def _decomposed(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        return la.matrix_decompose(self.matrix)
+
+    @property
+    def position(self) -> np.ndarray:
+        return self._decomposed[0]
+
+    @property
+    def rotation(self) -> np.ndarray:
+        return self._decomposed[1]
+
+    @property
+    def scale(self) -> np.ndarray:
+        return self._decomposed[2]
+
+    @property
+    def x(self) -> float:
+        return self.position[0]
+
+    @property
+    def y(self) -> float:
+        return self.position[1]
+
+    @property
+    def z(self) -> float:
+        return self.position[2]
+
+    @property
+    def scale_x(self) -> float:
+        return self.scale[0]
+
+    @property
+    def scale_y(self) -> float:
+        return self.scale[1]
+
+    @property
+    def scale_z(self) -> float:
+        return self.scale[2]
+
+    def __array__(self, dtype=None):
+        return self.matrix.astype(dtype)
+
+
+class AffineTransform(AffineBase):
+    """An affine tranformation"""
+
+    def __init__(
+        self,
+        matrix=None,
+        /,
+        *,
+        position=None,
+        rotation=None,
+        scale=None,
+        update_callback=None,
+    ) -> None:
+        super().__init__()
+        self.last_modified = perf_counter_ns()
+        self.update_callback = update_callback
+
+        if matrix is None:
+            matrix = np.eye(4, dtype=float)
+
+        self.untracked_matrix = np.asarray(matrix)
+
+        if position is not None:
+            self.position = position
+
+        if rotation is not None:
+            self.rotation = rotation
+
+        if scale is not None:
+            self.scale = scale
+
+    def flag_update(self):
+        self.last_modified = perf_counter_ns()
+
+        if self.update_callback is not None:
+            self.update_callback(self)
+
+    @property
+    def matrix(self):
+        view_array = self.untracked_matrix.view()
+        view_array.flags.writeable = False
+        return view_array
+
+    @matrix.setter
+    def matrix(self, value):
+        self.untracked_matrix[:] = value
+        self.flag_update()
+
+    @AffineBase.position.setter
+    def position(self, value):
+        self.untracked_matrix[:-1, -1] = value
+        self.flag_update()
+
+    @AffineBase.rotation.setter
+    def rotation(self, value):
+        self.matrix = la.matrix_make_transform(self.position, value, self.scale)
+
+    @AffineBase.scale.setter
+    def scale(self, value):
+        self.matrix = la.matrix_make_transform(self.position, self.rotation, value)
+
+    @AffineBase.x.setter
+    def x(self, value):
+        _, y, z = self.position
+        self.position = (value, y, z)
```

Comment:
Would it help if we made `obj.transform.position` a tuple, so it's readonly?

Reply:
> Would it help if we made obj.transform.position a tuple, so it's readonly?

Not really, because then a user could not leverage the numpy trickery and do e.g. `obj.transform.position += (1, 0, 0)`.

Reply:
Which the user can't leverage anyway, because this will also not update the underlying matrix from which the position originates :/

I will make this read-only for now and then we update once we find a better solution.